_posts/2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,253 @@
+</div>
+
+One aspect of program verification is that a single error might cover two or more places in the code. For example, post-conditions are asserted at every exit point of a declaration. Therefore, if this assertion fails, it fails on an exit point, and for the particular post-condition.
+[Figure 3a](#fig_previous_diagnostics) shows that the error was never reported on the post-condition itself before (top image). Following user feedback, on [Figure 3b](#fig_new_diagnostics) I made the error visible on post-conditions, using a red-orange squiggly line in-code &mdash; to indicate it is a secondary error &mdash; and using the regular error icon in the gutter.
+Moreover, because hovering is one of first way users can discover features, I added the following improvements to the hover experience, as shown in the bottom of [Figure 3b](#fig_new_diagnostics):
+
+* We quote the exact code portions that Dafny could not prove
+* On the <b style="color:#3794ff">Error</b> keyword, the usual color for links, I added a link to documentation we also wrote about how to [fix failing assertions](https://dafny.org/latest/DafnyRef/DafnyRef\#sec-verification-debugging).
+* "This is assertion #1 of 2" explains how many assertions there are and where is this assertion roughly located in verification, which proved to be useful for [splitting verification in smaller tasks using annotations](https://dafny.org/latest/DafnyRef/DafnyRef\#sec-assertion-batches).
+* "Resource usage: 9K RU" provides a deterministic metric on how many _resource units_ (RU) the solver used to reach this verification result. Resource units do not fluctuate like time would on multiple iterations, or when running on a different computer, although we found that both are correlated.
+* Hovering a _proved_ assertion reveals similar insights, with a positive message instead of an error, such as "Success: The divisior is never zero".
+* On hovering the line containing the method, if there are no other hovered assertions, I display a summary of verification as in [Figure 4](#fig_method_summary). This summary provides a link to the notion of [assertion batches]((https://dafny.org/latest/DafnyRef/DafnyRef\#sec-assertion-batches)) that helps splitting verification in smaller tasks, as well as a unicode warning icon `/!\` with a link to documentation to [fix verification when verification uses more than 10 million RU](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-verification-debugging-slow), a threshold I set by experience.
```

Comment:
By "line containing the method" did you mean "method containing the line"?

```suggestion
* On hovering over the method containing the line, if there are no other hovered assertions, I display a summary of verification as in [Figure 4](#fig_method_summary). This summary provides a link to the notion of [assertion batches]((https://dafny.org/latest/DafnyRef/DafnyRef\#sec-assertion-batches)) that helps splitting verification in smaller tasks, as well as a unicode warning icon `/!\` with a link to documentation to [fix verification when verification uses more than 10 million RU](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-verification-debugging-slow), a threshold I set by experience.
```

Reply:
> By "line containing the method" did you mean "method containing the line"?

It was actually "line containing the method name". Thanks for pointing it out!